File: job-agent-os/src/job_agent_os/schemas/application.py
"""Application schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ApplicationCreate(BaseModel):
    """Application creation request."""

    job_id: UUID
    resume_id: UUID
    priority: str = Field(default="medium", pattern="^(high|medium|low)$")
    notes: str | None = None


class ApplicationStatusUpdate(BaseModel):
    """Application status update request."""

    status: str
    stage: str | None = None
    notes: str | None = None
    next_follow_up: datetime | None = None


class ApplicationResponse(BaseModel):
    """Application response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    job_id: UUID
    resume_id: UUID
    status: str
    stage: str
    match_score: float | None = None
    match_report: dict | None = None
    priority: str
    notes: str | None = None
    applied_at: datetime | None = None
    next_follow_up: datetime | None = None
    created_at: datetime


class KanbanView(BaseModel):
    """Kanban board view."""

    pending: list[ApplicationResponse] = Field(default_factory=list)
    applied: list[ApplicationResponse] = Field(default_factory=list)
    written_test: list[ApplicationResponse] = Field(default_factory=list)
    round1: list[ApplicationResponse] = Field(default_factory=list)
    round2: list[ApplicationResponse] = Field(default_factory=list)
    hr_interview: list[ApplicationResponse] = Field(default_factory=list)
    offer: list[ApplicationResponse] = Field(default_factory=list)
    rejected: list[ApplicationResponse] = Field(default_factory=list)
    statistics: dict = Field(default_factory=dict)


class ApplicationStatistics(BaseModel):
    """Application statistics."""

    total_applications: int = 0
    by_status: dict = Field(default_factory=dict)
    by_platform: dict = Field(default_factory=dict)
    pass_rate: float = 0.0
    avg_match_score: float = 0.0
    stage_conversion: dict = Field(default_factory=dict)

File: job-agent-os/frontend/src/types/application.ts
/** Application types (matches backend schemas/application.py) */

export type ApplicationStatus =
  | "pending"
  | "applied"
  | "written_test"
  | "round1"
  | "round2"
  | "hr_interview"
  | "offer"
  | "rejected";

export interface ApplicationResponse {
  id: string;
  job_id: string;
  resume_id: string;
  status: ApplicationStatus;
  stage: string;
  match_score: number | null;
  match_report: Record<string, unknown> | null;
  priority: "high" | "medium" | "low" | string;
  notes: string | null;
  applied_at: string | null;
  next_follow_up: string | null;
  created_at: string;
}

export interface KanbanView {
  pending: ApplicationResponse[];
  applied: ApplicationResponse[];
  written_test: ApplicationResponse[];
  round1: ApplicationResponse[];
  round2: ApplicationResponse[];
  hr_interview: ApplicationResponse[];
  offer: ApplicationResponse[];
  rejected: ApplicationResponse[];
  statistics: Record<string, number>;
}

export interface ApplicationStatistics {
  total_applications: number;
  by_status: Record<string, number>;
  by_platform: Record<string, number>;
  pass_rate: number;
  avg_match_score: number;
  stage_conversion: Record<string, number>;
}

export interface ApplicationStatusUpdate {
  status: ApplicationStatus;
  stage?: string | null;
  notes?: string | null;
  next_follow_up?: string | null;
}

/** Column order for the kanban board (matches backend state machine). */
export const KANBAN_COLUMNS: { key: ApplicationStatus; label: string }[] = [
  { key: "pending", label: "待投递" },
  { key: "applied", label: "已投递" },
  { key: "written_test", label: "笔试" },
  { key: "round1", label: "一面" },
  { key: "round2", label: "二面" },
  { key: "hr_interview", label: "HR面" },
  { key: "offer", label: "Offer" },
  { key: "rejected", label: "已淘汰" },
];
